junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/NestedClassSelector.java

```diff
@@ -0,0 +1,117 @@
+ *
+ * <p>In this context, Java {@link Class} means anything that can be referenced
+ * as a {@link Class} on the JVM &mdash; for example, classes from other JVM
+ * languages such Groovy, Scala, etc.
+ *
+ * @since 1.6
+ * @see DiscoverySelectors#selectNestedClass(List, Class)
+ * @see DiscoverySelectors#selectNestedClass(List, String)
+ * @see org.junit.platform.engine.support.descriptor.ClassSource
+ */
+@API(status = STABLE, since = "1.6")
+public class NestedClassSelector implements DiscoverySelector {
+
+	private List<ClassSelector> enclosingClassSelectors;
+	private ClassSelector nestedClassSelector;
+
+	NestedClassSelector(List<String> enclosingClassNames, String nestedClassName) {
+		this.enclosingClassSelectors = enclosingClassNames.stream().map(ClassSelector::new).collect(toList());
+		this.nestedClassSelector = new ClassSelector(nestedClassName);
+	}
+
+	NestedClassSelector(List<Class<?>> enclosingClasses, Class<?> nestedClass) {
+		this.enclosingClassSelectors = enclosingClasses.stream().map(ClassSelector::new).collect(toList());
+		this.nestedClassSelector = new ClassSelector(nestedClass);
+	}
+
+	/**
+	 * Get the names of the classes enclosing the selected nested class.
+	 */
+	public List<String> getEnclosingClassNames() {
+		return enclosingClassSelectors.stream().map(ClassSelector::getClassName).collect(toList());
+	}
+
+	/**
+	 * Get the list of {@link Class} enclosing the selected nested
+	 * {@link Class}.
+	 *
+	 * <p>If the {@link Class} were not provided, but only the name of the
+	 * nested class and its enclosing classes, this method attempts to lazily
+	 * load the list of enclosing {@link Class} and throws a
+	 * {@link PreconditionViolationException} if the classes cannot be loaded.
+	 */
+	public List<Class<?>> getEnclosingClasses() {
+		return enclosingClassSelectors.stream().map(ClassSelector::getJavaClass).collect(toList());
+	}
+
+	/**
+	 * Get the name of the selected nested class.
+	 */
+	public String getNestedClassName() {
+		return nestedClassSelector.getClassName();
+	}
+
+	/**
+	 * Get the selected nested {@link Class}.
+	 *
+	 * <p>If the {@link Class} were not provided, but only the name of the
+	 * nested class and its enclosing classes, this method attempts to lazily
+	 * load the nested {@link Class} and throws a
+	 * {@link PreconditionViolationException} if the class cannot be loaded.
+	 */
+	public Class<?> getNestedClass() {
+		return nestedClassSelector.getJavaClass();
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) {
+			return true;
+		}
+		if (o == null || getClass() != o.getClass()) {
+			return false;
+		}
+		NestedClassSelector that = (NestedClassSelector) o;
+		return enclosingClassSelectors.equals(that.enclosingClassSelectors)
+				&& nestedClassSelector.equals(that.nestedClassSelector);
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(enclosingClassSelectors, nestedClassSelector);
+	}
+
```

Comment:
Please add `toString()` methods to both new selectors (see the existing selectors for examples).